src/mixins/createStatefulListenersMixin.ts

```diff
@@ -0,0 +1,158 @@
+import Promise from 'dojo-core/Promise';
+import WeakMap from 'dojo-core/WeakMap';
+
+export interface ActionRegistry {
+	get(id: string | symbol): Promise<Actionable<TargettedEventObject>>;
+	identify(value: Actionable<TargettedEventObject>): string | symbol;
+}
+
+export type ListenerOrArray = string | symbol | (string | symbol)[];
+
+export interface ListenersMap {
+	[type: string]: ListenerOrArray;
+}
+
+export interface StatefulListenersState {
+	listeners?: ListenersMap;
+}
+
+export interface StatefulListenersOptions<S extends StatefulListenersState> extends StatefulOptions<S> {
+	actionRegistry?: ActionRegistry;
+}
+
+export type StatefulListeners<S extends StatefulListenersState> = Stateful<S>;
+
+export interface StatefulListenersMixinFactory extends ComposeFactory<Stateful<StatefulListenersState>, StatefulListenersOptions<StatefulListenersState>> {
+	(options?: StatefulListenersOptions<StatefulListenersState>): Stateful<StatefulListenersState>;
+}
+
+function resolveListeners(registry: ActionRegistry, cache: Map<string | symbol, Actionable<TargettedEventObject>>, ref: ListenerOrArray): { value?: any; promise?: Promise<any>; } {
+	if (Array.isArray(ref)) {
+		const resolved = ref.map((item) => {
+			return resolveListeners(registry, cache, item);
+		});
+
+		let isSync = true;
+		const values: any[] = [];
+		for (const result of resolved) {
+			if (result.value) {
+				values.push(result.value);
+			} else {
+				isSync = false;
+				values.push(result.promise);
+			}
+		}
+
+		return isSync ? { value: values } : { promise: Promise.all(values) };
+	}
+
+	const id = <string | symbol> ref;
+	if (cache.has(id)) {
+		return { value: cache.get(id) };
+	}
+
+	const promise = registry.get(id);
+	promise.then((action) => {
+		cache.set(id, action);
+	});
+	return { promise };
+}
+
+interface ManagementState {
+	cache?: Map<string | symbol, Actionable<TargettedEventObject>>;
+	generation?: number;
+	handle?: Handle;
+	registry: ActionRegistry;
+}
+
+/**
+ * Map that holds state for manageListeners by widget instance.
+ */
+const managementMap = new WeakMap<StatefulListeners<StatefulListenersState>, ManagementState>();
+
+/**
+ * Internal statechange listener which replaces the listeners on the instance.
+ */
+function manageListeners(evt: StateChangeEvent<StatefulListenersState>): void {
+	const widget: StatefulListeners<StatefulListenersState> = <any> evt.target;
+
+	// Assume this function cannot be called without the widget being in the management map.
+	const state = managementMap.get(widget);
+	// Initialize cache.
+	if (!state.cache) {
+		state.cache = new Map<string | symbol, Actionable<TargettedEventObject>>();
+	}
+	// Increment the generation vector. Used when listeners are replaced asynchronously to ensure
+	// no newer state is overriden.
+	const generation = state.generation = (state.generation || 0) + 1;
+
+	// Unsetting listeners causes previous listeners to be removed.
+	const listeners = evt.state.listeners || {};
```

Comment:
Minor point, but should we be explicit about the type here?  We are depending upon TypeScript to contextually infer the type from the first value of the statement and the second value's type is overly broad.  TypeScript is "magically" saving us here.